common/src/main/java/org/apache/celeborn/common/network/server/MapDataPartition.java

```diff
@@ -0,0 +1,341 @@
+        storageFetcherPool.computeIfAbsent(
+            fileInfo.getMountPoint(),
+            k ->
+                Executors.newFixedThreadPool(
+                    threadsPerMountPoint,
+                    new ThreadFactoryBuilder()
+                        .setNameFormat(fileInfo.getMountPoint() + "-reader-thread-%d")
+                        .setUncaughtExceptionHandler(
+                            (t1, t2) -> {
+                              logger.warn("StorageFetcherPool thread:{}:{}", t1, t2);
+                            })
+                        .build()));
+    this.dataFileChanel = new FileInputStream(fileInfo.getFile()).getChannel();
+    this.indexChannel = new FileInputStream(fileInfo.getIndexPath()).getChannel();
+  }
+
+  private synchronized void updateLocalTarget(
+      long nTarget, long fileSize, long definedMinReadAheadMemory, long definedMaxReadAheadMemory) {
+    long target = nTarget;
+    if (target < definedMinReadAheadMemory) {
+      target = definedMinReadAheadMemory;
+    }
+    if (target > definedMaxReadAheadMemory) {
+      target = definedMaxReadAheadMemory;
+    }
+    if (target > fileSize) {
+      target = fileSize;
+    }
+    localBuffersTarget = (int) Math.ceil(target * 1.0 / fileInfo.getBufferSize());
+    localMemoryTarget = target;
+  }
+
+  public synchronized void setupDataPartitionReader(
+      int startSubIndex, int endSubIndex, long streamId, Channel channel) {
+    MapDataPartitionReader mapDataPartitionReader =
+        new MapDataPartitionReader(
+            startSubIndex,
+            endSubIndex,
+            fileInfo,
+            streamId,
+            channel,
+            () -> recycleStream.accept(streamId));
+    readers.put(streamId, mapDataPartitionReader);
+
+    // allocate resources when the first reader is registered
+    if (!bufferQueueInitialized) {
+      memoryManager.requestReadBuffers(
+          new ReadBufferRequest(
+              localBuffersTarget,
+              fileInfo.getBufferSize(),
+              (allocatedBuffers, throwable) -> onBuffer(allocatedBuffers)));
+      bufferQueueInitialized = true;
+    } else {
+      triggerRead();
+    }
+  }
+
+  // Read logic is executed on another thread.
+  public void onBuffer(List<ByteBuf> buffers) {
+    if (isReleased || bufferQueue.isReleased()) {
+      buffers.forEach(memoryManager::recycleReadBuffer);
+      return;
+    }
+
+    try {
+      bufferQueue.add(buffers);
+      inFlightBuffers -= buffers.size();
+    } catch (Exception e) {
+      // this branch means that this bufferQueue is closed
+      buffers.forEach(memoryManager::recycleReadBuffer);
+      return;
+    }
+
+    if (bufferQueue.size() > Math.min(localBuffersTarget / 2 + 1, readAheadMin)) {
+      triggerRead();
+    }
+  }
+
+  public void recycle(ByteBuf buffer) {
+    if (bufferQueue.numBuffersOccupied() > localBuffersTarget) {
+      bufferQueue.recycle(buffer);
+    } else {
+      buffer.clear();
+      bufferQueue.add(buffer);
```

Comment:
Remove BufferQueue.add method, and let bufferQueue.recycle to decide whether should add the buffer back to buffers or return to MemoryManager.